segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/ServiceConfig.java

```diff
@@ -49,7 +49,7 @@
     public static final Property<Integer> CACHE_POLICY_TARGET_UTILIZATION = Property.named("cacheTargetUtilizationPercent", (int) (100 * CachePolicy.DEFAULT_TARGET_UTILIZATION));
     public static final Property<Integer> CACHE_POLICY_MAX_UTILIZATION = Property.named("cacheMaxUtilizationPercent", (int) (100 * CachePolicy.DEFAULT_MAX_UTILIZATION));
     public static final Property<Integer> CACHE_POLICY_MAX_TIME = Property.named("cacheMaxTimeSeconds", 30 * 60);
-    public static final Property<Integer> CACHE_POLICY_GENERATION_TIME = Property.named("cacheGenerationTimeSeconds", 5);
+    public static final Property<Integer> CACHE_POLICY_GENERATION_TIME = Property.named("cacheGenerationTimeSeconds", 1);
```

Comment:
I don't fully understand the additional overhead that introduces when we drop from 5s to 1s. I'm also wondering whether this value isn't related to the cache size (i.e., larger cache can have a larger cache policy generation time while a smaller cache requires a shorter cache policy generation time), in which case, perhaps we could derive it from the cache size.

Could you elaborate, please?

Reply:
The overhead I'm speaking of consists of the work that the Cache Manager is doing. At each iteration it does the following steps:
1. Query all its registered clients of their state (how much data they have in the cache)
2. Determine if it needs to cleanup.
3. Instruct all registered clients to evict data progressively. This is done by incrementing the "oldest generation" that is allowed for cache entries until the clients report that enough data has been evicted to bring the utilization below the threshold.

A cache client here is anything that stores data in the cache (nothing to do with the Pravega Client):
1. Each Segment has a Cache Client in its Read Index (used for caching segment data)
2. Each Segment that uses Attributes has an additional Cache Client for its BTree nodes.
3. Each Table Segment has an additional Cache Client for storing hot keys.

The biggest and most frequently used chunk of the cache is used by 1), followed by 2) and then 3). 

The Cache Manager needs to manage all cache clients for all active Segments on the Segment store Instance (from all the containers). 

In the worst case, if we have 8 containers with max allowed segments (25K) we may have 200K segments so up to 400K clients. This is an extreme case and in this case each such segment will not have much data in the cache anyway, so the overhead will be in looping through all these clients.

The other extreme is where we have a few hot segments (such as our focused benchmark tests). We may only have a handful of clients, but the Segment Read Index and Attribute Index Caches are very active. The Segment Read Index will likely have hundreds of thousands of entries (assuming a 32KB average size within a 3GB cache); in this case querying is not expensive (we keep stats separately) but even evicting 1% of these items may include thousands of modifications to the index (and removals from the cache), all of which have to be done while holding a lock on the index. The Index is an optimized AVL tree so it's not the bottleneck, but the cache is.

My concern was that by increasing the frequency of cache cleanups, we may be spending more time maintaining these caches. However after further thinking, it is highly plausible that the increased frequency will result in smaller evictions so we are actually going to make the system better - instead of a large eviction every 5 seconds, we have much smaller ones every 1s.

Reply:
Let's talk specifically about these points:

> In the worst case, if we have 8 containers with max allowed segments (25K) we may have 200K segments so up to 400K clients.

> The other extreme is where we have a few hot segments (such as our focused benchmark tests). We may only have a handful of clients, but the Segment Read Index and Attribute Index Caches are very active.

For the first case, I understand that the time to "visit" each client is short, but I don't have a sense of how short. How long does it take to loop through 400k clients in the worst case? Is it more than 1s?

For the second case, the visit to each client is more costly, but I still don't understand how long each visit takes, roughly, I'm not necessarily asking about precise numbers.

Reply:
> How long does it take to loop through 400k clients in the worst case? Is it more than 1s?

No, we are not talking about seconds. All clients are kept in a list so we need to iterate through the list. A 400K iteration shouldn't take more than 1ms.

> For the second case, the visit to each client is more costly, but I still don't understand how long each visit takes, roughly, I'm not necessarily asking about precise numbers.

This varies by client. Let's consider each of them in turn:
- Segment Read Index. 
    - FetchStatus loops through all the generations available (the spread can be as small as 1 or as large as 2000 (one generation == one iteration of the Cache Manager; so max generation spread should be the value of cache.maxCacheTime (30 mins)). So a pretty quick one for each.    
    - UpdateGenerations (apply policy) is a bit tricker. It acquires the lock (one lock per Segment read index) and then loops through every single entry in the index (AVLTree in-order walk) and verifies if eligible for removal. If so, they are added to a removal list (we can't remove while iterating) and then they are removed from the index and from the cache. Everything is done under the lock.
        - UpdateGenerations is applied one generation at a time (the Cache Manager increments the generation, applies it, and then re-queries to see if it needs to more. So it will not hold the lock for too long. The biggest time will be spent removing items from the cache, but fortunately when we do #4044 this will be done in <50ns and there is no need to hold the lock while removing from the cache anymore.
- AttributeIndex and TableKeyCache (they are much simpler and smaller)
    - Both GetStatus and UpdateGenerations loop through all the cache entries and sum up their values or evict them, respectively. There are locks involved in both of them.

Reply:
No, they are Java `synchronized` blocks. I found that read-write locks are performing worse than `synchronized` blocks when locking for short amounts of time; they are intended for longer operations, such as when doing some sort of IO.

I don't have any monitoring for cycle times. We can add this as a metric, but I'll do it with a different issue. #4186.